mautic/mautic
# Segment dependency tree

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR adds new tab to the segment detail page for the segments that have some **Segment Membership** filters. Segments without this filter will not have this tab visible as it would not provide any value.

This is how it draws a tree chart for those segments that have some segment membership filters:
![Screen Shot 2021-09-07 at 09 39 29](https://user-images.githubusercontent.com/1235442/132304702-b7cb0c29-5562-4da7-ba31-7ecc07f64a07.png)

This is how the chart looks like if there are either duplicate segments in the tree structure or if there is a loop:
![Screen Shot 2021-09-07 at 11 08 32](https://user-images.githubusercontent.com/1235442/132317945-ab0d0854-63b9-43b5-9cef-959f5a801937.png)

This feature will not affect segment detail page load time. In fact, it will not load on the page load but it will load only when a user clicks on the tab. So no resources will be wasted to build this when not needed.

#### ToDo:
- [x] Test how it works for loops
- [x] Test how it works for duplicated segments

#### Steps to test this PR:
1. Open new Mautic instance with this change applied by clicking the GitPod button bellow.
2. Ensure the new tab is showing only on the right segments that have some segment membership filters.
3. Build a several-level deep segment dependency tree by including/excluding segments and confirm the tree chart is correct.
4. Test what happens if you create a segment loop. For example edit the segment at the end of the tree and include the segment at the top of the tree.
5. Test what happens if you duplicate including/excluding segments in different levels.

#### Other areas of Mautic that may be affected by the change:
1. Just segment detail page

[//]: # ( As applicable: )
#### List of areas covered by the unit and/or functional tests:
1. The functional test builds a 3-level segment dependency and ensures that the API endpoint returns the right data.

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11276"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
